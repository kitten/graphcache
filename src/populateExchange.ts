import {
  DocumentNode,
  buildClientSchema,
  visitWithTypeInfo,
  TypeInfo,
  FragmentDefinitionNode,
  GraphQLSchema,
  IntrospectionQuery,
  FragmentSpreadNode,
  NameNode,
  ASTNode,
  isCompositeType,
  isAbstractType,
  Kind,
  visit,
} from 'graphql';

import { pipe, tap, map } from 'wonka';
import { Exchange, Operation } from 'urql/core';

import { getName, getSelectionSet, unwrapType } from './ast';
import { makeDict } from './store';
import { invariant, warn } from './helpers/help';

interface PopulateExchangeOpts {
  schema: IntrospectionQuery;
}

/** An exchange for auto-populating mutations with a required response body. */
export const populateExchange = ({
  schema: ogSchema,
}: PopulateExchangeOpts): Exchange => ({ forward }) => {
  const schema = buildClientSchema(ogSchema);
  /** List of operation keys that have already been parsed. */
  const parsedOperations = new Set<number>();
  /** List of operation keys that have not been torn down. */
  const activeOperations = new Set<number>();
  /** Collection of fragments used by the user. */
  const userFragments: UserFragmentMap = makeDict();
  /** Collection of actively in use type fragments. */
  const activeTypeFragments: TypeFragmentMap = makeDict();

  /** Handle mutation and inject selections + fragments. */
  const handleIncomingMutation = (op: Operation) => {
    if (op.operationName !== 'mutation') {
      return op;
    }

    const activeSelections: TypeFragmentMap = makeDict();
    for (const name in activeTypeFragments) {
      activeSelections[name] = activeTypeFragments[name].filter(s =>
        activeOperations.has(s.key)
      );
    }

    return {
      ...op,
      query: addFragmentsToQuery(
        schema,
        op.query,
        activeSelections,
        userFragments
      ),
    };
  };

  /** Handle query and extract fragments. */
  const handleIncomingQuery = ({ key, operationName, query }: Operation) => {
    if (operationName !== 'query') {
      return;
    }

    activeOperations.add(key);
    if (parsedOperations.has(key)) {
      return;
    }

    parsedOperations.add(key);

    const [extractedFragments, newFragments] = extractSelectionsFromQuery(
      schema,
      query
    );

    for (let i = 0, l = extractedFragments.length; i < l; i++) {
      const fragment = extractedFragments[i];
      userFragments[getName(fragment)] = fragment;
    }

    for (let i = 0, l = newFragments.length; i < l; i++) {
      const fragment = newFragments[i];
      const type = getName(fragment.typeCondition);
      const current =
        activeTypeFragments[type] || (activeTypeFragments[type] = []);

      (fragment as any).name.value += current.length;
      current.push({ key, fragment });
    }
  };

  const handleIncomingTeardown = ({ key, operationName }: Operation) => {
    if (operationName === 'teardown') {
      activeOperations.delete(key);
    }
  };

  return ops$ => {
    return pipe(
      ops$,
      tap(handleIncomingQuery),
      tap(handleIncomingTeardown),
      map(handleIncomingMutation),
      forward
    );
  };
};

type UserFragmentMap<T extends string = string> = Record<
  T,
  FragmentDefinitionNode
>;

type TypeFragmentMap<T extends string = string> = Record<T, TypeFragment[]>;

interface TypeFragment {
  /** Operation key where selection set is being used. */
  key: number;
  /** Selection set. */
  fragment: FragmentDefinitionNode;
}

/** Gets typed selection sets and fragments from query */
export const extractSelectionsFromQuery = (
  schema: GraphQLSchema,
  query: DocumentNode
) => {
  const extractedFragments: FragmentDefinitionNode[] = [];
  const newFragments: FragmentDefinitionNode[] = [];
  const typeInfo = new TypeInfo(schema);

  visit(
    query,
    visitWithTypeInfo(typeInfo, {
      Field: node => {
        if (node.selectionSet) {
          const type = getTypeName(typeInfo);
          newFragments.push({
            kind: Kind.FRAGMENT_DEFINITION,
            typeCondition: {
              kind: Kind.NAMED_TYPE,
              name: nameNode(type),
            },
            name: nameNode(`${type}_PopulateFragment_`),
            selectionSet: node.selectionSet,
          });
        }
      },
      FragmentDefinition: node => {
        extractedFragments.push(node);
      },
    })
  );

  return [extractedFragments, newFragments];
};

/** Replaces populate decorator with fragment spreads + fragments. */
export const addFragmentsToQuery = (
  schema: GraphQLSchema,
  query: DocumentNode,
  activeTypeFragments: TypeFragmentMap,
  userFragments: UserFragmentMap
) => {
  const typeInfo = new TypeInfo(schema);

  const requiredUserFragments: Record<
    string,
    FragmentDefinitionNode
  > = makeDict();

  const additionalFragments: Record<
    string,
    FragmentDefinitionNode
  > = makeDict();

  /** Fragments provided and used by the current query */
  const existingFragmentsForQuery: Set<string> = new Set();

  return visit(
    query,
    visitWithTypeInfo(typeInfo, {
      Field: {
        enter: node => {
          if (!node.directives) {
            return;
          }

          const directives = node.directives.filter(
            d => getName(d) !== 'populate'
          );
          if (directives.length === node.directives.length) {
            return;
          }

          const possibleTypes = getTypes(schema, typeInfo);
          const newSelections = possibleTypes.reduce((p, possibleType) => {
            const typeFrags = activeTypeFragments[possibleType.name];
            if (!typeFrags) {
              return p;
            }

            for (let i = 0, l = typeFrags.length; i < l; i++) {
              const { fragment } = typeFrags[i];
              const fragmentName = getName(fragment);
              const usedFragments = getUsedFragments(fragment);

              // Add used fragment for insertion at Document node
              for (let j = 0, l = usedFragments.length; j < l; j++) {
                const name = usedFragments[j];
                if (!existingFragmentsForQuery.has(name)) {
                  requiredUserFragments[name] = userFragments[name];
                }
              }

              // Add fragment for insertion at Document node
              additionalFragments[fragmentName] = fragment;

              p.push({
                kind: Kind.FRAGMENT_SPREAD,
                name: nameNode(fragmentName),
              });
            }

            return p;
          }, [] as FragmentSpreadNode[]);

          const existingSelections = getSelectionSet(node);

          const selections =
            existingSelections.length + newSelections.length !== 0
              ? [...newSelections, ...existingSelections]
              : [
                  {
                    kind: Kind.FIELD,
                    name: nameNode('__typename'),
                  },
                ];

          return {
            ...node,
            directives,
            selectionSet: {
              kind: Kind.SELECTION_SET,
              selections,
            },
          };
        },
      },
      Document: {
        enter: node => {
          node.definitions.reduce((set, definition) => {
            if (definition.kind === 'FragmentDefinition') {
              set.add(definition.name.value);
            }
            return set;
          }, existingFragmentsForQuery);
        },
        leave: node => {
          const definitions = [...node.definitions];
          for (const key in additionalFragments)
            definitions.push(additionalFragments[key]);
          for (const key in requiredUserFragments)
            definitions.push(requiredUserFragments[key]);
          return { ...node, definitions };
        },
      },
    })
  );
};

const nameNode = (value: string): NameNode => ({
  kind: Kind.NAME,
  value,
});

/** Get all possible types for node with TypeInfo. */
const getTypes = (schema: GraphQLSchema, typeInfo: TypeInfo) => {
  const type = unwrapType(typeInfo.getType());
  if (!isCompositeType(type)) {
    warn(
      'Invalid type: The type ` + type + ` is used with @populate but does not exist.',
      17
    );
    return [];
  }

  return isAbstractType(type) ? schema.getPossibleTypes(type) : [type];
};

/** Get name of non-abstract type for adding to 'activeTypeFragments'. */
const getTypeName = (typeInfo: TypeInfo) => {
  const type = unwrapType(typeInfo.getType());
  invariant(
    type && !isAbstractType(type),
    'Invalid TypeInfo state: Found no flat schema type when one was expected.',
    18
  );

  return type.toString();
};

/** Get fragment names referenced by node. */
const getUsedFragments = (node: ASTNode) => {
  const names: string[] = [];

  visit(node, {
    FragmentSpread: f => {
      names.push(getName(f));
    },
  });

  return names;
};
