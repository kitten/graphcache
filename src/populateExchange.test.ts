import {
  buildSchema,
  print,
  introspectionFromSchema,
  visit,
  DocumentNode,
  ASTKindToNode,
} from 'graphql';

import gql from 'graphql-tag';
import { fromValue, pipe, fromArray, toArray } from 'wonka';
import { Client, Operation } from 'urql/core';

import { populateExchange } from './populateExchange';

const schemaDef = `
  interface Node {
    id: ID!
  }

  type User implements Node {
    id: ID!
    name: String!
    age: Int!
    todos: [Todo]
  }

  type Todo implements Node {
    id: ID!
    text: String!
    creator: User!
  }

  union UnionType = User | Todo

  type Query {
    todos: [Todo!]
    users: [User!]!
  }

  type Mutation {
    addTodo: [Todo]
    removeTodo: [Node]
    updateTodo: [UnionType]
  }
`;

const getNodesByType = <T extends keyof ASTKindToNode, N = ASTKindToNode[T]>(
  query: DocumentNode,
  type: T
) => {
  let result: N[] = [];

  visit(query, {
    [type]: n => {
      result = [...result, n];
    },
  });
  return result;
};

const schema = introspectionFromSchema(buildSchema(schemaDef));

beforeEach(jest.clearAllMocks);

const exchangeArgs = {
  forward: a => a as any,
  client: {} as Client,
};

describe('on mutation', () => {
  const operation = {
    key: 1234,
    operationName: 'mutation',
    query: gql`
      mutation MyMutation {
        addTodo @populate
      }
    `,
  } as Operation;

  describe('mutation query', () => {
    it('matches snapshot', async () => {
      const response = pipe<Operation, any, Operation[]>(
        fromValue(operation),
        populateExchange({ schema })(exchangeArgs),
        toArray
      );
      expect(print(response[0].query)).toMatchInlineSnapshot(`
        "mutation MyMutation {
          addTodo {
            __typename
          }
        }
        "
      `);
    });
  });
});

describe('on query -> mutation', () => {
  const queryOp = {
    key: 1234,
    operationName: 'query',
    query: gql`
      query {
        todos {
          id
          text
          creator {
            id
            name
          }
        }
        users {
          todos {
            text
          }
        }
      }
    `,
  } as Operation;

  const mutationOp = {
    key: 5678,
    operationName: 'mutation',
    query: gql`
      mutation MyMutation {
        addTodo @populate
      }
    `,
  } as Operation;

  describe('mutation query', () => {
    it('matches snapshot', async () => {
      const response = pipe<Operation, any, Operation[]>(
        fromArray([queryOp, mutationOp]),
        populateExchange({ schema })(exchangeArgs),
        toArray
      );

      expect(print(response[1].query)).toMatchInlineSnapshot(`
        "mutation MyMutation {
          addTodo {
            ...Todo_PopulateFragment_0
            ...Todo_PopulateFragment_1
          }
        }

        fragment Todo_PopulateFragment_0 on Todo {
          id
          text
          creator {
            id
            name
          }
        }

        fragment Todo_PopulateFragment_1 on Todo {
          text
        }
        "
      `);
    });
  });
});

describe('on (query w/ fragment) -> mutation', () => {
  const queryOp = {
    key: 1234,
    operationName: 'query',
    query: gql`
      query {
        todos {
          ...TodoFragment
          creator {
            ...CreatorFragment
          }
        }
      }

      fragment TodoFragment on Todo {
        id
        text
      }

      fragment CreatorFragment on User {
        id
        name
      }
    `,
  } as Operation;

  const mutationOp = {
    key: 5678,
    operationName: 'mutation',
    query: gql`
      mutation MyMutation {
        addTodo @populate {
          ...TodoFragment
        }
      }

      fragment TodoFragment on Todo {
        id
        text
      }
    `,
  } as Operation;

  describe('mutation query', () => {
    it('matches snapshot', async () => {
      const response = pipe<Operation, any, Operation[]>(
        fromArray([queryOp, mutationOp]),
        populateExchange({ schema })(exchangeArgs),
        toArray
      );

      expect(print(response[1].query)).toMatchInlineSnapshot(`
        "mutation MyMutation {
          addTodo {
            ...Todo_PopulateFragment_0
            ...TodoFragment
          }
        }

        fragment TodoFragment on Todo {
          id
          text
        }

        fragment Todo_PopulateFragment_0 on Todo {
          ...TodoFragment
          creator {
            ...CreatorFragment
          }
        }

        fragment CreatorFragment on User {
          id
          name
        }
        "
      `);
    });

    it('includes user fragment', () => {
      const response = pipe<Operation, any, Operation[]>(
        fromArray([queryOp, mutationOp]),
        populateExchange({ schema })(exchangeArgs),
        toArray
      );

      const fragments = getNodesByType(response[1].query, 'FragmentDefinition');
      expect(
        fragments.filter(f => f.name.value === 'TodoFragment')
      ).toHaveLength(1);
    });
  });
});

describe('on (query w/ unused fragment) -> mutation', () => {
  const queryOp = {
    key: 1234,
    operationName: 'query',
    query: gql`
      query {
        todos {
          id
          text
        }
        users {
          ...UserFragment
        }
      }

      fragment UserFragment on User {
        id
        name
      }
    `,
  } as Operation;

  const mutationOp = {
    key: 5678,
    operationName: 'mutation',
    query: gql`
      mutation MyMutation {
        addTodo @populate
      }
    `,
  } as Operation;

  describe('mutation query', () => {
    it('matches snapshot', async () => {
      const response = pipe<Operation, any, Operation[]>(
        fromArray([queryOp, mutationOp]),
        populateExchange({ schema })(exchangeArgs),
        toArray
      );

      expect(print(response[1].query)).toMatchInlineSnapshot(`
        "mutation MyMutation {
          addTodo {
            ...Todo_PopulateFragment_0
          }
        }

        fragment Todo_PopulateFragment_0 on Todo {
          id
          text
        }
        "
      `);
    });

    it('excludes user fragment', () => {
      const response = pipe<Operation, any, Operation[]>(
        fromArray([queryOp, mutationOp]),
        populateExchange({ schema })(exchangeArgs),
        toArray
      );

      const fragments = getNodesByType(response[1].query, 'FragmentDefinition');
      expect(
        fragments.filter(f => f.name.value === 'UserFragment')
      ).toHaveLength(0);
    });
  });
});

describe('on query -> (mutation w/ interface return type)', () => {
  const queryOp = {
    key: 1234,
    operationName: 'query',
    query: gql`
      query {
        todos {
          id
          name
        }
        users {
          id
          text
        }
      }
    `,
  } as Operation;

  const mutationOp = {
    key: 5678,
    operationName: 'mutation',
    query: gql`
      mutation MyMutation {
        removeTodo @populate
      }
    `,
  } as Operation;

  describe('mutation query', () => {
    it('matches snapshot', async () => {
      const response = pipe<Operation, any, Operation[]>(
        fromArray([queryOp, mutationOp]),
        populateExchange({ schema })(exchangeArgs),
        toArray
      );

      expect(print(response[1].query)).toMatchInlineSnapshot(`
        "mutation MyMutation {
          removeTodo {
            ...Todo_PopulateFragment_0
            ...User_PopulateFragment_0
          }
        }

        fragment Todo_PopulateFragment_0 on Todo {
          id
          name
        }

        fragment User_PopulateFragment_0 on User {
          id
          text
        }
        "
      `);
    });
  });
});

describe('on query -> (mutation w/ union return type)', () => {
  const queryOp = {
    key: 1234,
    operationName: 'query',
    query: gql`
      query {
        todos {
          id
          name
        }
        users {
          id
          text
        }
      }
    `,
  } as Operation;

  const mutationOp = {
    key: 5678,
    operationName: 'mutation',
    query: gql`
      mutation MyMutation {
        updateTodo @populate
      }
    `,
  } as Operation;

  describe('mutation query', () => {
    it('matches snapshot', async () => {
      const response = pipe<Operation, any, Operation[]>(
        fromArray([queryOp, mutationOp]),
        populateExchange({ schema })(exchangeArgs),
        toArray
      );

      expect(print(response[1].query)).toMatchInlineSnapshot(`
        "mutation MyMutation {
          updateTodo {
            ...User_PopulateFragment_0
            ...Todo_PopulateFragment_0
          }
        }

        fragment User_PopulateFragment_0 on User {
          id
          text
        }

        fragment Todo_PopulateFragment_0 on Todo {
          id
          name
        }
        "
      `);
    });
  });
});

describe('on query -> teardown -> mutation', () => {
  const queryOp = {
    key: 1234,
    operationName: 'query',
    query: gql`
      query {
        todos {
          id
          text
        }
      }
    `,
  } as Operation;

  const teardownOp = {
    key: queryOp.key,
    operationName: 'teardown',
  } as Operation;

  const mutationOp = {
    key: 5678,
    operationName: 'mutation',
    query: gql`
      mutation MyMutation {
        addTodo @populate
      }
    `,
  } as Operation;

  describe('mutation query', () => {
    it('matches snapshot', async () => {
      const response = pipe<Operation, any, Operation[]>(
        fromArray([queryOp, teardownOp, mutationOp]),
        populateExchange({ schema })(exchangeArgs),
        toArray
      );

      expect(print(response[2].query)).toMatchInlineSnapshot(`
        "mutation MyMutation {
          addTodo {
            __typename
          }
        }
        "
      `);
    });

    it('only requests __typename', () => {
      const response = pipe<Operation, any, Operation[]>(
        fromArray([queryOp, teardownOp, mutationOp]),
        populateExchange({ schema })(exchangeArgs),
        toArray
      );
      getNodesByType(response[2].query, 'Field').forEach(field => {
        expect(field.name.value).toMatch(/addTodo|__typename/);
      });
    });
  });
});
