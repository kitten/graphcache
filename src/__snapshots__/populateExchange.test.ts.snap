// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on (query w/ fragment) -> mutation mutation query matches snapshot 1`] = `
"mutation MyMutation {
  addTodo {
    ...Todo_PopulateFragment_0
  }
}

fragment Todo_PopulateFragment_0 on Todo {
  ...TodoFragment
}

fragment TodoFragment on Todo {
  id
  text
}
"
`;

exports[`on (query w/ unused fragment) -> mutation mutation query matches snapshot 1`] = `
"mutation MyMutation {
  addTodo {
    ...Todo_PopulateFragment_0
  }
}

fragment Todo_PopulateFragment_0 on Todo {
  id
  text
}
"
`;

exports[`on mutation mutation query matches snapshot 1`] = `
"mutation MyMutation {
  addTodo {
    __typename
  }
}
"
`;

exports[`on query -> (mutation w/ interface return type) mutation query matches snapshot 1`] = `
"mutation MyMutation {
  removeTodo {
    ...Todo_PopulateFragment_0
    ...User_PopulateFragment_0
  }
}

fragment Todo_PopulateFragment_0 on Todo {
  id
  name
}

fragment User_PopulateFragment_0 on User {
  id
  text
}
"
`;

exports[`on query -> (mutation w/ union return type) mutation query matches snapshot 1`] = `
"mutation MyMutation {
  updateTodo {
    ...User_PopulateFragment_0
    ...Todo_PopulateFragment_0
  }
}

fragment User_PopulateFragment_0 on User {
  id
  text
}

fragment Todo_PopulateFragment_0 on Todo {
  id
  name
}
"
`;

exports[`on query -> mutation mutation query matches snapshot 1`] = `
"mutation MyMutation {
  addTodo {
    ...Todo_PopulateFragment_0
    ...Todo_PopulateFragment_1
  }
}

fragment Todo_PopulateFragment_0 on Todo {
  id
  text
  creator {
    id
    name
  }
}

fragment Todo_PopulateFragment_1 on Todo {
  text
}
"
`;

exports[`on query -> teardown -> mutation mutation query matches snapshot 1`] = `
"mutation MyMutation {
  addTodo {
    __typename
  }
}
"
`;
